src/rust/service/memcache/src/lib.rs

```diff
@@ -0,0 +1,167 @@
+use std::sync::Arc;
+use session::Session;
+use protocol_common::*;
+use protocol_memcache::*;
+
+pub enum RecvError {
+    Invalid,
+    Incomplete,
+    NoPending,
+    Pending,
+    Unknown,
+}
+
+impl From<ParseError> for RecvError {
+    fn from(other: ParseError) -> Self {
+        match other {
+            ParseError::Invalid => Self::Invalid,
+            ParseError::Incomplete => Self::Incomplete,
+            ParseError::Unknown => Self::Unknown,
+        }
+    }
+}
+
+pub enum SendError {
+    NoPending,
+    Pending,
+}
+
+pub struct Server {
+    parser: RequestParser,
+    pending: Option<Arc<Request>>,
+}
+
+impl Server {
```

Comment:
@thinkingfish - take a look and see if you think this is headed in the right direction.

Reply:
I checked the C implementation. It didn't try to declare per-command metrics only [generic ones](https://github.com/twitter/pelikan/blob/master/src/protocol/data/memcache/parse.h), but instead provided more detailed command metrics in the `server/data/process.[ch]` [files](https://github.com/twitter/pelikan/blob/master/src/server/twemcache/data/process.h#L39)

This would mean client/server/proxy will each declare their own metrics which can be tedious, but it allows the metric name to be succinct in each context.

Reply:
I still want to do this somewhere below the core/server, core/proxy so that they compose in from some library. I think the idea of having a "service" library and moving maybe all of the request/response stats into there would make sense. I'll take another pass and see if I can get it to something closer to my end goal. Might be able to (ab)use features and macros to get us succinct names and keep the necessary logic in the "service" libraries.